### PR TITLE
Support insert into a table with row filter and hidden columns

### PR DESCRIPTION
**Pull request by kokosing**
Support insert into a table with row filter and hidden columns

### CODE REVIEW

**Comment by kokosing** [comment]
Fixes #10005

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 1]
```diff
@@ -486,4 +495,64 @@ public void testInsert()
...
+        assertions.query("SELECT * FROM mock.tiny.nation_with_hidden_column")
+                .assertThat()
+                .skippingTypesCheck()
+                .matches("VALUES (BIGINT '0', 'ALGERIA', BIGINT '0', ' haggle. carefully final deposits detect slyly agai')");
+        // TODO https://github.com/trinodb/trino/issues/10005 - support insert into a table with hidden columns
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_hidden_column VALUES (101, 'POLAND', 0, 'No comment')"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("scope: 5, fields mappings: 4");
+        // TODO https://github.com/trinodb/trino/issues/10005 - support insert into a table with hidden columns
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_hidden_column VALUES (0, 'POLAND', 0, 'No comment')"))
```
Comment (gaurav8297) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java]: This case is repeated

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 2]
```diff
@@ -486,4 +495,64 @@ public void testInsert()
...
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("scope: 5, fields mappings: 4");
+        assertThatThrownBy(() -> assertions.query("UPDATE mock.tiny.nation_with_hidden_column SET name = 'POLAND'"))
+                .isInstanceOf(TrinoException.class)
+                .hasMessageContaining("Updating a table with a row filter is not supported");
+        assertions.query("DELETE FROM mock.tiny.nation_with_hidden_column WHERE regionkey < 5")
```
Comment (gaurav8297) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:522]: Can we also test deleting from the table with a hidden column condition?

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 3]
```diff
@@ -486,4 +495,64 @@ public void testInsert()
...
+        assertions.query("SELECT count(*) FROM mock.tiny.nation_with_hidden_column")
+                .assertThat()
+                .skippingTypesCheck()
+                .matches("VALUES BIGINT '25'");
+        // TODO https://github.com/trinodb/trino/issues/10006 - support insert into a table with row filter that is using hidden columns
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_hidden_column VALUES (101, 'POLAND', 0, 'No comment')"))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("scope: 5, fields mappings: 4");
+        // TODO https://github.com/trinodb/trino/issues/10006 - support insert into a table with row filter that is using hidden columns
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_hidden_column VALUES (0, 'POLAND', 0, 'No comment')"))
```
Comment (gaurav8297) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:546]: Repeated case

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 4]
```diff
@@ -541,12 +540,12 @@ public void testRowFilterOnHiddenColumn()
                 .matches("VALUES BIGINT '25'");
         // TODO https://github.com/trinodb/trino/issues/10006 - support insert into a table with row filter that is using hidden columns
         assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_hidden_column VALUES (101, 'POLAND', 0, 'No comment')"))
-                .isInstanceOf(IllegalArgumentException.class)
-                .hasMessage("scope: 5, fields mappings: 4");
+                .isInstanceOf(ArrayIndexOutOfBoundsException.class)
+                .hasMessage("Index 4 out of bounds for length 4");
```
Comment (gaurav8297) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:544]: Maybe we can improve the error message from the user perspective?

**Review by ksobolew — APPROVED** [review]
LGTM % @gaurav8297's comments

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java (hunk 4 above)
Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:544]: It is way harder to fix. I would need to rewrite a lot of planer code for insert. Same for error message.

**Review by kokosing — COMMENTED** [review]
ac

**Review by gaurav8297 — APPROVED** [review]

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java (hunk 3 above)
Comment (kokosing) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java]: `nationkey` is irrelevant. Row filter here is not using `nationkey` column at all.

Comment (ksobolew) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java]: Ah, indeed. But the one in `testRowFilterWithHiddenColumns` was also removed and it wasn't repeated?

Comment (ksobolew) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java]: Or am I imagining things, hm

Comment (ksobolew) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java]: > Or am I imagining things, hm

Looks like I am, sorry about that :)

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 5]
```diff
@@ -467,7 +467,19 @@ private RelationPlan getInsertPlan(
...
+                failIfPredicateIsNotMeet(metadata, session, PERMISSION_DENIED, AccessDeniedException.PREFIX + "Cannot insert into a table restricted with a row filter"),
+                node -> {
+                    Scope accessControlScope = analysis.getAccessControlScope(table);
+                    // hidden fields are not accessible in insert
+                    return Scope.builder()
+                            .like(accessControlScope)
+                            .withRelationType(
+                                    accessControlScope.getRelationId(),
+                                    new RelationType(accessControlScope.getRelationType().getAllFields().stream()
+                                            .filter(field -> !field.isHidden())
```
[the remark this answers is not in the document]
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:479]: How about this `RelationType#withOnlyVisibleFields`

**Comment by kokosing** [comment]
@Praveen2112 AC, PTAL

**Review by Praveen2112 — APPROVED** [review]

**Pull request merged**